File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/VocService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.controller.dto.voc.VocRequest
import pt.isel.odin.model.Voc

/**
 * Service for vocs.
 */
interface VocService {

    /**
     * Gets a voc by its id.
     *
     * @param id the voc id.
     */
    fun getById(id: Long): Voc

    /**
     * Gets all vocs.
     */
    fun getAll(): List<Voc>

    /**
     * Saves a voc.
     *
     * @param vocRequest the voc info to save.
     * @param email the User.
     */
    fun save(vocRequest: Voc, email: String): Voc

    /**
     * Updates a voc.
     *
     * @param vocRequest the voc info to update.
     */
    fun update(vocRequest: VocRequest): Voc

    /**
     * Deletes a voc by its id.
     *
     * @param id the voc id.
     */
    fun delete(id: Long)

    /**
     * Gets all vocs by User.
     *
     * @param email the User.
     */
    fun getByStudent(email: String): List<Voc>
}

File: code/frontend/Tyr/src/services/VocService.ts
import {makeApiRequest} from '../axios/apiRequest';
import {Voc, VocRequest} from "../model/Voc";

/**
 * Service to manage vocs
 */
export class VocService {

    static basePath = '/vocs';

    static async getById(id: number): Promise<Voc> {
        return makeApiRequest('get', `${(VocService.basePath)}/${id}`);
    }

    static async getAll(): Promise<Voc[]> {
        return makeApiRequest('get', VocService.basePath);
    }

    static async save(vocRequest: VocRequest): Promise<Voc> {
        return makeApiRequest('post', `${(VocService.basePath)}/save`, vocRequest);
    }

    static async update(vocRequest: VocRequest): Promise<Voc> {
        return makeApiRequest('put', `${(VocService.basePath)}/update`, vocRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(VocService.basePath)}/${id}`);
    }

    static async getVocsByUser(): Promise<GetVocsByUserResponse> {
        return makeApiRequest('get', `${(VocService.basePath)}/user`);
    }
}

interface GetVocsByUserResponse {
    vocs: Voc[];
}